OfficeDev/office-ui-fabric-react
# Fix ContextualMenuItemButton's secondary text never read by screen-reader

#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000

#### Description of changes

#### Overview

We received an accessibility report in #6360:

> Bug 5 : [Accessibility][Screen reader]: Secondary text is not announced when user navigates in context menu under ContextualMenu with icons and secondary text.

After debugging, I found that this is because we set an `aria-label` on `ContextualMenuButton` equal to either `props.ariaLabel` or `item.text || item.name || ''`. The item's secondary text is never included in that `aria-label` as `item.secondaryText` and therefore never read.

#### The Change

Initially, I thought the fix was to either conditionally concatenate primary and secondary text to produce `aria-label` or replace `aria-label` with `aria-labelledby` referencing both primary and secondary text nodes via unique IDs. The spec recommends `aria-labelledby` for multiple nodes with text visible.

Characteristic | Value
-- | --
Related Concepts: | A related concept is label in XForms [XForms] and HTML [HTML].
Used in Roles: | All elements of the base markup
**Value:** | **ID reference list**

However, I found that I could remove the usage of `aria-label` completely (unless specified by the user via props) and the bug was fixed. Both primary and secondary text were then announced by Narrator because they are present in the DOM as child text nodes. [See Narrator Buddy Output](#narrator-buddy-output).

#### Conclusion

I'm recommending this change, because it seems to be backwards compatible for devs that are already supplying their own `ariaLabel` or relying on text contents `item.text` or `item.name` be read by assistive technology.

I believe that the most likely usage of `ariaLabel` in a `ContextualMenuItemButton` is when the item's text  is unclear such as `"X"`. In this case, the existing code's default value will be `"X"` for `aria-label` and not any more useful. This would require a developer to supply their own `ariaLabel` value anyway.

#### Narrator Buddy Output

**Before**

Default ContextualMenu Example

```shell
Office UI Fabric React Examples - Microsoft Edge window, Office UI Fabric React Examples, menu, 
New, menu item,
Rename, menu item,
Edit, menu item,
Properties, menu item,
Link same window, menu item,
Tooltip, http://bing.com/.
Link new window, menu item,
Link click, menu item,
```

ContextualMenu w/ Secondary Text Example

<pre>
Office UI Fabric React Examples - Microsoft Edge window, Office UI Fabric React Examples, menu, 
<b>Later Today, menu item,
Tomorrow, menu item,
This Weekend, menu item,
Next Week, menu item,</b>
</pre>

**After**

Default ContextualMenu Example

```shell
Office UI Fabric React Examples - Microsoft Edge window, Office UI Fabric React Examples, menu, 
New, menu item,
Rename, menu item,
Edit, menu item,
Properties, menu item,
Link same window, menu item,
Tooltip, http://bing.com/.
Link new window, menu item,
Link click, menu item,
```
ContextualMenu w/ Secondary Text Example

<pre>
Office UI Fabric React Examples - Microsoft Edge window, Office UI Fabric React Examples, menu, 
<b>Later Today 7:00 PM, menu item,
Tomorrow Thu. 8:00 AM, menu item,
This Weekend Sat. 10:00 AM, menu item,
Next Week Mon. 8:00 AM, menu item,</b>
</pre>

#### Focus areas to test

1. Navigate to the ContextualMenu examples page
1. Enable Narrator
1. Open the various ContextualMenus on the page, each item should have its text read aloud.
1. Open the ContextualMenu example with secondary text.
1. Narrator should read the primary and secondary text of each button that is focused within the menu.

**References**
- https://www.w3.org/WAI/PF/aria-1.1/states_and_properties#aria-labelledby

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/OfficeDev/office-ui-fabric-react/pull/6670)
